majority-multisign/src/MajorityMultiSign/OnChain.hs

```diff
@@ -69,19 +69,26 @@ hasNewSignatures (UpdateKeysAct keys) MajorityMultiSignDatum {signers} ctx =
 {-# INLINEABLE hasCorrectToken #-}
 hasCorrectToken :: MajorityMultiSignValidatorParams -> ScriptContext -> MajorityMultiSignDatum -> Bool
 hasCorrectToken MajorityMultiSignValidatorParams {asset} ctx expectedDatum =
-  traceIfFalse "Couldn't find asset" (isJust assetTxOut)
-    && traceIfFalse
-      "Incorrect output datum"
-      ((assetTxOut >>= txOutDatumHash) == findDatumHash (Datum $ PlutusTx.toBuiltinData expectedDatum) (scriptContextTxInfo ctx))
+  isJust result
   where
     continuing :: [TxOut]
     continuing = Ledger.getContinuingOutputs ctx
 
     checkAsset :: TxOut -> Maybe TxOut
     checkAsset txOut = if assetClassValueOf (txOutValue txOut) asset > 0 then Just txOut else Nothing
 
-    assetTxOut :: Maybe TxOut
-    assetTxOut = firstJust checkAsset continuing
+    traceIfNothing :: BuiltinString -> Maybe a -> Maybe a
+    traceIfNothing errMsg = maybe (trace errMsg Nothing) Just
+
+    result :: Maybe ()
+    result = do
+      !assetTxOut <- traceIfNothing "Couldn't find asset" $ firstJust checkAsset continuing
+      let !datumHash = traceIfNothing "Continuing output does not have datum" $ txOutDatumHash assetTxOut
+          !expectedDatumHash =
+            traceIfNothing "Datum map does not have expectedDatum" $
+              findDatumHash (Datum $ PlutusTx.toBuiltinData expectedDatum) (scriptContextTxInfo ctx)
+      !_ <- traceIfFalse "Incorrect output datum" <$> liftA2 (==) datumHash expectedDatumHash
```

Comment:
Why was this strictness required? Would be nicce to have been able to use `void` here, but it wouldnt follow through with strictness.

Reply:
Oh damn, I missed this comment. #plutus-learning chat wasn't sure at first if PlutusTx strict or lazy by default, so I made sure that `traceIfNothing` gets called with bangs. They are not needed, actually, according to Las. He said that PlutusTx *is* strict